loader/loader.go

```diff
@@ -1188,110 +1188,246 @@ func fetchMatchedLiteral(ctx *tcontext.Context, router *router.Table, schema, ta
 	return targetSchema, targetTable
 }
 
-func (l *Loader) restoreData(ctx context.Context) error {
-	begin := time.Now()
+// `restore Schema Job` present a data structure of schema restoring job
+type restoreSchemaJob struct {
+	loader   *Loader
+	session  *DBConn
+	database string // database name
+	table    string // table name, empty if it's a schema of database
+	filepath string // file path of dumpped schema file
+}
 
-	baseConn, err := l.toDB.GetBaseConn(ctx)
-	if err != nil {
-		return err
+// `jobQueue` of schema restoring which (only)support consumptions concurrently
+type jobQueue struct {
+	ctx           context.Context
+	msgq          chan *restoreSchemaJob // job message queue channel
+	consumerCount uint16                 // count of consumers
+	eg            *errgroup.Group        // err wait group of consumer's go-routines
+}
+
+// `newJobQueue` consturct a jobQueue
+func newJobQueue(ctx context.Context, consumerCount uint16) *jobQueue {
+	eg, selfCtx := errgroup.WithContext(ctx)
+	return &jobQueue{
+		ctx:           selfCtx,
+		msgq:          make(chan *restoreSchemaJob, 1),
+		consumerCount: consumerCount,
+		eg:            eg,
 	}
-	defer func() {
-		err2 := l.toDB.CloseBaseConn(baseConn)
-		if err2 != nil {
-			l.logger.Warn("fail to close connection", zap.Error(err2))
-		}
-	}()
+}
 
-	dbConn := &DBConn{
-		cfg:      l.cfg,
-		baseConn: baseConn,
-		resetBaseConnFn: func(*tcontext.Context, *conn.BaseConn) (*conn.BaseConn, error) {
-			return nil, terror.ErrDBBadConn.Generate("bad connection error restoreData")
-		},
+// `push` will append a job to the queue
+func (q *jobQueue) push(job *restoreSchemaJob) error {
+	var err error
+	select {
+	case <-q.ctx.Done():
+		err = q.ctx.Err()
+	case q.msgq <- job:
 	}
+	return err
+}
 
-	dispatchMap := make(map[string]*fileJob)
+// `close` wait jobs done and close queue forever
+func (q *jobQueue) close() error {
+	// queue is closing
```

Comment:
we might let `jobQueue` use a child context, and let this `close` cancel that context, so we don't need `handle` the only job in `msgq`.

(and I think one capacity for `msgq` will cause a frequent goroutine switch. it's not a big problem)

Reply:
Don't worry, `jobQueue.eg` will propagate cancel signal when error occurs. Procedure  of consumer handler also watching the `context.Done()` signal.